### ISSUE TITLE
Properly polymorphic `withRunInIO`.

### ISSUE THREAD

**Issue by effectfully** [fpco/unliftio#11]
Why not make `withRunInIO` properly polymorphic? There is no impredicativity involved:

```haskell
withRunInIO :: MonadUnliftIO m => ((forall a. m a -> IO a) -> IO b) -> m b
withRunInIO inner = withUnliftIO $ \u -> inner (unliftIO u)
```

Then

```haskell
race :: MonadUnliftIO m => m a -> m b -> m (Either a b)
race a b = withRunInIO $ \runInIO -> A.race (runInIO a) (runInIO b)
```

**Comment by snoyberg** [comment]
That's a great catch! Assuming this is backwards compatible with all of the supported GHCs (Travis should tell us that), I don't see a problem. You interested in sending a PR?

**Comment by effectfully** [comment]
Sure, let me concont one.

**Comment by snoyberg** [comment]
Implemented in #12, thanks!
